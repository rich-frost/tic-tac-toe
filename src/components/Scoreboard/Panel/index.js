import React from 'react';
import PropTypes from 'prop-types';
import './Panel.css';

export default function Panel({ label, value }) {
    return (
        <div className="scoreboard-panel">
            <h4>{label}</h4>
            <div className="scoreboard-panel-value">{value}</div>
        </div>
    );
}

Panel.propTypes = {
    label: PropTypes.string.isRequired,
    value: PropTypes.number.isRequired,
};
