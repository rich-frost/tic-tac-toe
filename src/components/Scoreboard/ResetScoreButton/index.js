import React from 'react';
import PropTypes from 'prop-types';

export default function ResetScoreButton({ setScore }) {
    return (
        <button onClick={() => setScore({ human: 0, computer: 0, draw: 0 })}>Reset score</button>
    );
}

ResetScoreButton.propTypes = {
    setScore: PropTypes.func.isRequired,
};
