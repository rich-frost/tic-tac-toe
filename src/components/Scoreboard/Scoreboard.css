.scoreboard {
    margin-right: 1rem;
    width: 300px;
    padding: 1rem;
}
