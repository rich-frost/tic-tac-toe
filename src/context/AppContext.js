import React, { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

export const AppContext = createContext();

export const AppContextProvider = ({ children }) => {
    const [score, setScore] = useState({ human: 0, computer: 0, draw: 0 });
    const [playerName, setPlayerName] = useState('Player 1');

    return (
        <AppContext.Provider value={{ score, setScore, playerName, setPlayerName }}>
            {children}
        </AppContext.Provider>
    );
};

AppContextProvider.propTypes = {
    children: PropTypes.element.isRequired,
};

export const useAppContext = () => useContext(AppContext);
