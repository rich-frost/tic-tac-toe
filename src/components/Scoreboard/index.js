import React, { useMemo } from 'react';
import { useAppContext } from '../../context/AppContext';
import './Scoreboard.css';
import Panel from './Panel';
import ResetScoreButton from './ResetScoreButton';

export default function Scoreboard() {
    const { score, setScore, playerName } = useAppContext();
    const labelMappings = useMemo(
        () => ({
            human: playerName,
            computer: 'Computer',
            draw: 'Draws',
        }),
        [playerName]
    );
    const hasScore = useMemo(() => {
        return Object.values(score).some((item) => item > 0);
    }, [score]);
    return (
        <div className="scoreboard panel">
            <h3>SCORE BOARD</h3>
            {Object.entries(score).map(([label, value]) => (
                <Panel key={label} {...{ label: labelMappings[label], value }} />
            ))}
            {hasScore && <ResetScoreButton setScore={setScore} />}
        </div>
    );
}
