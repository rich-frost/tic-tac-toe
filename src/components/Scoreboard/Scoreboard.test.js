import React from 'react';
import { render } from '@testing-library/react';
import Scoreboard from './';
import { useAppContext } from '../../context/AppContext';

jest.mock('../../context/AppContext');

test('renders default scoreboard', () => {
    useAppContext.mockImplementation(() => ({
        score: { human: 0, computer: 0, draw: 0 },
        setScore: () => {},
        playerName: 'Richard',
    }));
    const { container } = render(<Scoreboard />);
    expect(container).toMatchSnapshot();
});

test('renders scoreboard with values and reset button', () => {
    useAppContext.mockImplementation(() => ({
        score: { human: 1, computer: 3, draw: 6 },
        setScore: () => {},
        playerName: 'Rich',
    }));
    const { container } = render(<Scoreboard />);
    expect(container).toMatchSnapshot();
});
