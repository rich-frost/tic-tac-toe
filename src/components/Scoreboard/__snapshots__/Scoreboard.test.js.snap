// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default scoreboard 1`] = `
<div>
  <div
    class="scoreboard panel"
  >
    <h3>
      SCORE BOARD
    </h3>
    <div
      class="scoreboard-panel"
    >
      <h4>
        Richard
      </h4>
      <div
        class="scoreboard-panel-value"
      >
        0
      </div>
    </div>
    <div
      class="scoreboard-panel"
    >
      <h4>
        Computer
      </h4>
      <div
        class="scoreboard-panel-value"
      >
        0
      </div>
    </div>
    <div
      class="scoreboard-panel"
    >
      <h4>
        Draws
      </h4>
      <div
        class="scoreboard-panel-value"
      >
        0
      </div>
    </div>
  </div>
</div>
`;

exports[`renders scoreboard with values and reset button 1`] = `
<div>
  <div
    class="scoreboard panel"
  >
    <h3>
      SCORE BOARD
    </h3>
    <div
      class="scoreboard-panel"
    >
      <h4>
        Rich
      </h4>
      <div
        class="scoreboard-panel-value"
      >
        1
      </div>
    </div>
    <div
      class="scoreboard-panel"
    >
      <h4>
        Computer
      </h4>
      <div
        class="scoreboard-panel-value"
      >
        3
      </div>
    </div>
    <div
      class="scoreboard-panel"
    >
      <h4>
        Draws
      </h4>
      <div
        class="scoreboard-panel-value"
      >
        6
      </div>
    </div>
    <button>
      Reset score
    </button>
  </div>
</div>
`;
